File: android/app/src/main/java/noah/moohvie/viewmodels/QuizViewModel.kt
package noah.moohvie.viewmodels

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import noah.moohvie.models.AnswerOption
import noah.moohvie.models.Question
import noah.moohvie.models.QuestionBank
import noah.moohvie.models.QuestionPriority
import noah.moohvie.models.QuizLength

class QuizViewModel : ViewModel() {
    var currentQuestionIndex: Int by mutableStateOf(0)
        private set
    var isFinished: Boolean by mutableStateOf(false)
        private set

    private var selectedGenres: Set<Int> = emptySet()
    private var excludedGenres: Set<Int> = emptySet()

    var maxRuntime: Int? by mutableStateOf(null)
        private set
    var minVoteAverage: Double? by mutableStateOf(null)
        private set
    var minReleaseYear: Int? by mutableStateOf(null)
        private set
    var maxReleaseYear: Int? by mutableStateOf(null)
        private set

    var questions: List<Question> by mutableStateOf(emptyList())
        private set

    fun start(length: QuizLength) {
        questions = when (length) {
            QuizLength.SHORT -> QuestionBank.questions.filter { it.priority == QuestionPriority.ESSENTIAL }
            QuizLength.MEDIUM -> QuestionBank.questions.filter {
                it.priority == QuestionPriority.ESSENTIAL || it.priority == QuestionPriority.IMPORTANT
            }
            QuizLength.LONG -> QuestionBank.questions
        }
        reset()
    }

    val currentQuestion: Question
        get() = questions[currentQuestionIndex]

    val progress: Float
        get() = if (questions.isEmpty()) 0f else currentQuestionIndex.toFloat() / questions.size

    fun selectAnswer(option: AnswerOption) {
        selectedGenres = selectedGenres + option.genresToAdd
        excludedGenres = excludedGenres + option.genresToExclude

        option.maxRuntime?.let { maxRuntime = minOf(maxRuntime ?: it, it) }
        option.minVoteAverage?.let { minVoteAverage = maxOf(minVoteAverage ?: it, it) }
        option.minReleaseYear?.let { minReleaseYear = maxOf(minReleaseYear ?: it, it) }
        option.maxReleaseYear?.let { maxReleaseYear = minOf(maxReleaseYear ?: it, it) }

        if (currentQuestionIndex < questions.size - 1) {
            currentQuestionIndex += 1
        } else {
            isFinished = true
        }
    }

    val finalGenres: List<Int>
        get() = (selectedGenres - excludedGenres).toList()

    fun reset() {
        currentQuestionIndex = 0
        selectedGenres = emptySet()
        excludedGenres = emptySet()
        maxRuntime = null
        minVoteAverage = null
        minReleaseYear = null
        maxReleaseYear = null
        isFinished = false
    }
}

File: Moohvie/Services/quizviewmodel.swift
import Foundation
import Combine

enum QuizLength: Int {
    case short = 5
    case medium = 12
    case long = 20
}

@MainActor
class QuizViewModel: ObservableObject {
    @Published var currentQuestionIndex = 0
    @Published var selectedGenres: Set<Int> = []
    @Published var excludedGenres: Set<Int> = []
    @Published var isFinished = false

    @Published var maxRuntime: Int? = nil
    @Published var minVoteAverage: Double? = nil
    @Published var minReleaseYear: Int? = nil
    @Published var maxReleaseYear: Int? = nil

    var questions: [Question] = []

    func start(length: QuizLength) {
        switch length {
        case .short:
            questions = QuestionBank.questions.filter { $0.priority == .essential }
        case .medium:
            questions = QuestionBank.questions.filter { $0.priority == .essential || $0.priority == .important }
        case .long:
            questions = QuestionBank.questions
        }
        reset()
    }

    var currentQuestion: Question {
        questions[currentQuestionIndex]
    }

    var progress: Double {
        guard !questions.isEmpty else { return 0 }
        return Double(currentQuestionIndex) / Double(questions.count)
    }

    func selectAnswer(_ option: AnswerOption) {
        selectedGenres.formUnion(option.genresToAdd)
        excludedGenres.formUnion(option.genresToExclude)

        if let runtime = option.maxRuntime {
            maxRuntime = min(maxRuntime ?? runtime, runtime)
        }
        if let vote = option.minVoteAverage {
            minVoteAverage = max(minVoteAverage ?? vote, vote)
        }
        if let minYear = option.minReleaseYear {
            minReleaseYear = max(minReleaseYear ?? minYear, minYear)
        }
        if let maxYear = option.maxReleaseYear {
            maxReleaseYear = min(maxReleaseYear ?? maxYear, maxYear)
        }

        if currentQuestionIndex < questions.count - 1 {
            currentQuestionIndex += 1
        } else {
            isFinished = true
        }
    }

    var finalGenres: [Int] {
        Array(selectedGenres.subtracting(excludedGenres))
    }

    var finalExcludedGenres: [Int] {
        Array(excludedGenres.subtracting(selectedGenres))
    }

    func applyMood(_ preset: MoodPreset) {
        selectedGenres = Set(preset.genreIDs)
        excludedGenres = []
    }

    func reset() {
        currentQuestionIndex = 0
        selectedGenres = []
        excludedGenres = []
        maxRuntime = nil
        minVoteAverage = nil
        minReleaseYear = nil
        maxReleaseYear = nil
        isFinished = false
    }
}
